<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Use sitePath to find fixation and parallel sites • sitePath</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Use sitePath to find fixation and parallel sites">
<meta property="og:description" content="sitePath">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sitePath</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/sitePath.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wuaipinglab/sitePath/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Use sitePath to find fixation and parallel sites</h1>
                        <h4 data-toc-skip class="author">Chengyang Ji</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/wuaipinglab/sitePath/blob/HEAD/vignettes/sitePath.Rmd" class="external-link"><code>vignettes/sitePath.Rmd</code></a></small>
      <div class="hidden name"><code>sitePath.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>In viral evolution, fixed substitutions in the nucleic acid or protein level are closely associated with maintaining viral function, while parallel mutation reflects the competitive nature in adaptive selection. The continued accumulation of large-scale viral sequence data enhances the challenge of identifying important mutations in a quick and accurate manner. In sitePath, the phylogenetic tree was separated into a set of inheritable phylogenetic pathways via an automated pathway-division method. Then, for each phylogenetic pathway, the identification of fixed substitutions was transformed into a local-optimal-solution problem directed by a minimal entropy algorithm. Finally, the parallel mutation was determined based on the recurrence of mutations among at least two phylogenetic pathways.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>sitePath</code> package is made for the high-throughput identification of fixed substitutions and parallel mutations in viruses from a single phylogenetic tree. This is achieved by three major steps:</p>
<ol style="list-style-type: decimal">
<li>Clustering phylogenetic terminals</li>
<li>Identifying phylogenetic pathways</li>
<li>Finding fixed and parallel mutations</li>
</ol>
</div>
<div class="section level2">
<h2 id="clustering-phylogenetic-terminals">Clustering phylogenetic terminals<a class="anchor" aria-label="anchor" href="#clustering-phylogenetic-terminals"></a>
</h2>
<p>The firs step is to import phylogenetic tree and multiple sequence alignment files. For now, <code>sitePath</code> accepts <code>phylo</code> object and <code>alignment</code> object. Functions from <code>ggtree</code> and <code>seqinr</code> are able to handle most file formats.</p>
<div class="section level3">
<h3 id="import-tree-file">Import tree file<a class="anchor" aria-label="anchor" href="#import-tree-file"></a>
</h3>
<p>The S3 <code>phylo</code> class is a common data structure for phylogenetic analysis in R. The CRAN package <a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link">ape</a> provides basic parsing function for reading tree files. The Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/treeio.html" class="external-link">treeio</a> provides more comprehensive parsing utilities.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wuaipinglab.github.io/sitePath/" class="external-link">sitePath</a></span><span class="op">)</span>

<span class="va">tree_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ZIKV.newick"</span>, package <span class="op">=</span> <span class="st">"sitePath"</span><span class="op">)</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="va">tree_file</span><span class="op">)</span></code></pre></div>
<p>It is highly recommended that the file stores a rooted tree as R would consider the tree is rooted by default and re-rooting the tree in R is difficult. Also, we expect the tree to have no super long branches. A bad example is shown below:</p>
<p><img src="sitePath_files/figure-html/bad_tree_example-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="import-sequence-alignment-file">Import sequence alignment file<a class="anchor" aria-label="anchor" href="#import-sequence-alignment-file"></a>
</h3>
<p>Most multiple sequence alignment format can be parsed by <a href="https://cran.r-project.org/web/packages/seqinr/index.html" class="external-link">seqinr</a>. There is a wrapper function for parsing and adding the sequence alignment. Set “cl.cores” in <code>options</code> to the number of cores you want to use for multiprocessing.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alignment_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ZIKV.fasta"</span>, package <span class="op">=</span> <span class="st">"sitePath"</span><span class="op">)</span>

<span class="co"># options(list("cl.cores" = 10))</span>

<span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addMSA.html">addMSA</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">alignment_file</span>, <span class="st">"fasta"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="clustering-using-site-polymorphism">Clustering using site polymorphism<a class="anchor" aria-label="anchor" href="#clustering-using-site-polymorphism"></a>
</h3>
<p>The <code>addMSA</code> function will match the sequence names in tree and alignment. Also, the function uses polymorphism of each site to cluster sequences for identifying phylogenetic pathways.</p>
</div>
</div>
<div class="section level2">
<h2 id="identifying-phylogenetic-pathways">Identifying phylogenetic pathways<a class="anchor" aria-label="anchor" href="#identifying-phylogenetic-pathways"></a>
</h2>
<p>After importing the tree and sequence file, <code>sitePath</code> is ready to identify phylogenetic pathways.</p>
<div class="section level3">
<h3 id="the-impact-of-threshold-on-resolving-lineages">The impact of threshold on resolving lineages<a class="anchor" aria-label="anchor" href="#the-impact-of-threshold-on-resolving-lineages"></a>
</h3>
<p>The impact of threshold depends on the tree topology hence there is no universal choice. The function <code>sneakPeak</code> samples thresholds and calculates the resulting number of paths. <em>The use of this function can help choose the threshold.</em></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">preassessment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lineagePath.html">sneakPeek</a></span><span class="op">(</span><span class="va">paths</span>, makePlot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/sneakPeek_plot-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="choose-a-threshold">Choose a threshold<a class="anchor" aria-label="anchor" href="#choose-a-threshold"></a>
</h3>
<p>The default threshold is the first ‘stable’ value to produce the same number of phylogenetic pathways. You can directly use the return of <code>addMSA</code> if you want the default or choose other threshold by using function <code>lineagePath</code>. The choice of the threshold really depends. Here 18 is used as an example.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lineagePath.html">lineagePath</a></span><span class="op">(</span><span class="va">preassessment</span>, <span class="fl">18</span><span class="op">)</span>
<span class="va">paths</span>
<span class="co">#&gt; This is a 'lineagePath' object.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 4 lineage paths using 18 as "major SNP" threshold</span></code></pre></div>
<p>You can visualize the result.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_paths-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="finding-fixed-and-parallel-mutations">Finding fixed and parallel mutations<a class="anchor" aria-label="anchor" href="#finding-fixed-and-parallel-mutations"></a>
</h2>
<p>Now you’re ready to find fixation and parallel mutations.</p>
<div class="section level3">
<h3 id="entropy-minimization">Entropy minimization<a class="anchor" aria-label="anchor" href="#entropy-minimization"></a>
</h3>
<p>The <code>sitesMinEntropy</code> function perform entropy minimization on every site for each lineage. The fixation and parallel mutations can be derived from the function’s return value.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">minEntropy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sitesMinEntropy.html">sitesMinEntropy</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fixation-mutations">Fixation mutations<a class="anchor" aria-label="anchor" href="#fixation-mutations"></a>
</h3>
<p>The hierarchical search is done by <code>fixationSites</code> function. The function detects the site with fixation mutation.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fixations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixationSites.html">fixationSites</a></span><span class="op">(</span><span class="va">minEntropy</span><span class="op">)</span>
<span class="va">fixations</span>
<span class="co">#&gt; This is a 'fixationSites' object.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Result for 4 paths:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 139 894 2074 2086 2634 3045 988 1143 2842 3398 107 1118 3353 </span>
<span class="co">#&gt; No reference sequence specified. Using alignment numbering</span></code></pre></div>
<p>To get the position of all the resulting sites, <code>allSitesName</code> can be used on the return of <code>fixationSites</code> and also other functions like <code>SNPsites</code> and <code>parallelSites</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">allSites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/allSitesName.html">allSitesName</a></span><span class="op">(</span><span class="va">fixations</span><span class="op">)</span>
<span class="va">allSites</span>
<span class="co">#&gt;  [1] "139"  "894"  "2074" "2086" "2634" "3045" "988"  "1143" "2842" "3398"</span>
<span class="co">#&gt; [11] "107"  "1118" "3353"</span></code></pre></div>
<p>If you want to retrieve the result of a single site, you can pass the result of <code>fixationSites</code> and the site index to <code>extractSite</code> function. The output is a <code>sitePath</code> object which stores the tip names.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractSite.html">extractSite</a></span><span class="op">(</span><span class="va">fixations</span>, <span class="fl">139</span><span class="op">)</span></code></pre></div>
<p>It is also possible to retrieve the tips involved in the fixation of the site.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/extractTips.html">extractTips</a></span><span class="op">(</span><span class="va">fixations</span>, <span class="fl">139</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt;  [1] "ANK57896" "AMD61711" "AQS26698" "APG56458" "AUI42289" "AMR39834"</span>
<span class="co">#&gt;  [7] "AWH65848" "APO08504" "AMX81917" "AVZ47169" "AMX81916" "AMD61710"</span>
<span class="co">#&gt; [13] "AMK49492" "AMX81915" "AOC50652" "APH11611" "BBC70847" "AUF35022"</span>
<span class="co">#&gt; [19] "ATL14618" "AUF35021" "AVV62004" "BAX00477"</span>
<span class="co">#&gt; attr(,"AA")</span>
<span class="co">#&gt; [1] "S"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt;   [1] "BAV89190"     "AOI20067"     "AMM43325"     "AMM43326"     "AUI42329"    </span>
<span class="co">#&gt;   [6] "AUI42330"     "ANC90425"     "AMT75536"     "ANF16414"     "AMR68932"    </span>
<span class="co">#&gt;  [11] "ANA12599"     "AMM39806"     "AMR39830"     "AMV49165"     "AMO03410"    </span>
<span class="co">#&gt;  [16] "ANO46307"     "AVG19275"     "ANN44857"     "ANO46306"     "ANO46309"    </span>
<span class="co">#&gt;  [21] "ANO46305"     "ANO46303"     "ARB08102"     "ANO46302"     "AHZ13508"    </span>
<span class="co">#&gt;  [26] "ANO46304"     "ANO46301"     "ANO46308"     "AOG18296"     "AOO19564"    </span>
<span class="co">#&gt;  [31] "AUI42194"     "APC60215"     "AMQ48986"     "ATG29307"     "ART29828"    </span>
<span class="co">#&gt;  [36] "AWF93617"     "ATG29284"     "ATG29287"     "ATG29303"     "AWF93619"    </span>
<span class="co">#&gt;  [41] "AWF93618"     "AQM74762"     "AUD54964"     "AQM74761"     "ATG29306"    </span>
<span class="co">#&gt;  [46] "ASL68974"     "ATG29267"     "ASL68978"     "AQX32985"     "ATG29315"    </span>
<span class="co">#&gt;  [51] "AQZ41956"     "ARI68105"     "ASU55505"     "AQZ41949"     "ASL68979"    </span>
<span class="co">#&gt;  [56] "ATG29299"     "ATI21641"     "ATG29270"     "ATG29291"     "AOY08536"    </span>
<span class="co">#&gt;  [61] "ANO46297"     "ANO46298"     "AQZ41950"     "AQZ41951"     "ARU07183"    </span>
<span class="co">#&gt;  [66] "ANG09399"     "AQZ41954"     "AOY08533"     "AQZ41947"     "AQZ41948"    </span>
<span class="co">#&gt;  [71] "ATG29292"     "ATG29295"     "AOW32303"     "AVZ25033"     "AOC50654"    </span>
<span class="co">#&gt;  [76] "AQZ41953"     "ATG29301"     "ATG29276"     "APO08503"     "AMC13913"    </span>
<span class="co">#&gt;  [81] "AMC13912"     "APO39243"     "APO39229"     "AQZ41952"     "AQZ41955"    </span>
<span class="co">#&gt;  [86] "AMK49165"     "ARB07976"     "APB03018"     "AMC13911"     "APB03019"    </span>
<span class="co">#&gt;  [91] "ASU55416"     "ANK57897"     "AWH65849"     "AMZ03556"     "ASU55417"    </span>
<span class="co">#&gt;  [96] "ANW07476"     "APY24199"     "AMA12086"     "AMH87239"     "APY24198"    </span>
<span class="co">#&gt; [101] "APO36913"     "ALX35659"     "AOG18295"     "ANQ92019"     "AML81028"    </span>
<span class="co">#&gt; [106] "APY24200"     "AMD16557"     "ARU07074"     "AOX49264"     "AOX49265"    </span>
<span class="co">#&gt; [111] "AOY08518"     "ARB07962"     "AMX81919"     "AMM39805"     "ARX97119"    </span>
<span class="co">#&gt; [116] "AMB37295"     "AMK79468"     "AML82110"     "AMR39831"     "AMX81918"    </span>
<span class="co">#&gt; [121] "ANC90426"     "ALU33341"     "ASB32509"     "AMA12085"     "AMU04506"    </span>
<span class="co">#&gt; [126] "AMA12087"     "AMA12084"     "AQU12485"     "AMS00611"     "AMQ48981"    </span>
<span class="co">#&gt; [131] "AOY08538"     "APH11492"     "AOY08517"     "AOY08541"     "AOO54270"    </span>
<span class="co">#&gt; [136] "AND01116"     "ARU07076"     "AMK49164"     "APG56457"     "AOR82892"    </span>
<span class="co">#&gt; [141] "ATB53752"     "ANH10698"     "AOR82893"     "ARU07075"     "AMB18850"    </span>
<span class="co">#&gt; [146] "YP_009428568" "AMQ48982"     "ART29823"     "APW84876"     "ASK51714"    </span>
<span class="co">#&gt; [151] "ARB07953"     "APW84872"     "AOY08525"     "APW84873"     "AOY08535"    </span>
<span class="co">#&gt; [156] "AVZ25035"     "ARB07932"     "AOY08523"     "AOY08542"     "ASW34087"    </span>
<span class="co">#&gt; [161] "AOY08537"     "APB03020"     "ART29826"     "ART29825"     "AOS90220"    </span>
<span class="co">#&gt; [166] "AMN14620"     "APW84874"     "APW84875"     "BAV82373"     "AOS90221"    </span>
<span class="co">#&gt; [171] "AOS90224"     "APB03021"     "APO39232"     "AOS90223"     "APO39237"    </span>
<span class="co">#&gt; [176] "ANH22038"     "APW84877"     "APO39236"     "AOY08546"     "AOY08516"    </span>
<span class="co">#&gt; [181] "APO39233"     "AOS90222"     "AOO53981"     "AOY08521"     "AOO85388"    </span>
<span class="co">#&gt; [186] "APO39228"     "ARB07967"     "ANF04752"     "AOE22997"     "APQ41782"    </span>
<span class="co">#&gt; [191] "APQ41786"     "ASU55393"     "ASU55404"     "ASU55423"     "ANB66182"    </span>
<span class="co">#&gt; [196] "ASU55425"     "ASU55420"     "AQX32986"     "ASU55422"     "APQ41784"    </span>
<span class="co">#&gt; [201] "ANC90428"     "ASU55415"     "ASU55418"     "ARM59239"     "ASU55408"    </span>
<span class="co">#&gt; [206] "ASU55424"     "ASU55390"     "ASU55419"     "ASU55391"     "AMM39804"    </span>
<span class="co">#&gt; [211] "ASU55411"     "ANB66183"     "ASU55421"     "AMZ03557"     "ASU55392"    </span>
<span class="co">#&gt; [216] "AQX32987"     "ASU55403"     "ASU55399"     "APQ41783"     "ANS60026"    </span>
<span class="co">#&gt; [221] "ANB66184"     "ASU55426"     "ASU55412"     "ASU55413"     "ASU55410"    </span>
<span class="co">#&gt; [226] "ASU55397"     "ASU55400"     "ASU55409"     "APB03017"     "ASU55395"    </span>
<span class="co">#&gt; [231] "ASU55396"     "AOY08524"     "ASU55394"     "ASU55414"     "ASU55405"    </span>
<span class="co">#&gt; [236] "AMC33116"     "ASU55406"     "ASU55398"     "ASU55407"     "AMQ34003"    </span>
<span class="co">#&gt; [241] "AMQ34004"     "ASU55401"     "ASU55402"    </span>
<span class="co">#&gt; attr(,"AA")</span>
<span class="co">#&gt; [1] "N"</span></code></pre></div>
<p>Use <code>plot</code> on a <code>sitePath</code> object to visualize the fixation mutation of a single site. Alternatively, use <code>plotSingleSite</code> on an <code>fixationSites</code> object with the site specified.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_sitePath-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">fixations</span>, <span class="fl">139</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_sitePath-2.png" width="700"></p>
<p>To have an overall view of the transition of fixation mutation:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fixations</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_fixations-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="parallel-mutations">Parallel mutations<a class="anchor" aria-label="anchor" href="#parallel-mutations"></a>
</h3>
<p>Parallel mutation can be found by the <code>parallelSites</code> function. There are four ways of defining the parallel mutation: <code>all</code>, <code>exact</code>, <code>pre</code> and <code>post</code>. Here <code>exact</code> is used as an example.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">paraSites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parallelSites.html">parallelSites</a></span><span class="op">(</span><span class="va">minEntropy</span>, minSNP <span class="op">=</span> <span class="fl">1</span>, mutMode <span class="op">=</span> <span class="st">"exact"</span><span class="op">)</span>
<span class="va">paraSites</span>
<span class="co">#&gt; This is a 'parallelSites' object.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Result for 4 paths:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 105 1264 1226 1717 988 2611 2787 2749 3328 3162 1857 3046 1016 1171 1327 3076 106 2357 573 1404 940 1180 </span>
<span class="co">#&gt; No reference sequence specified. Using alignment numbering</span></code></pre></div>
<p>The result of a single site can be visualized by <code>plotSingleSite</code> function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paraSites</span>, <span class="fl">105</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>To have an overall view of the parallel mutations:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">paraSites</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_parallel-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="additional-functions">Additional functions<a class="anchor" aria-label="anchor" href="#additional-functions"></a>
</h2>
<p>This part is extra and experimental but might be useful when pre-assessing your data. We’ll use an example to demonstrate.</p>
<div class="section level3">
<h3 id="inspect-one-site">Inspect one site<a class="anchor" aria-label="anchor" href="#inspect-one-site"></a>
</h3>
<p>The <code>plotSingleSite</code> function will color the tree according to amino acids if you use the output of <code>lineagePath</code> function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="fl">139</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_sites-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="fl">763</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/plot_sites-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="snp-sites">SNP sites<a class="anchor" aria-label="anchor" href="#snp-sites"></a>
</h3>
<p>An SNP site could potentially undergo fixation event. The <code>SNPsites</code> function predicts possible SNP sites and the result could be what you’ll expect to be fixation mutation. Also, a tree plot with mutation could be visualized with <code>plotMutSites</code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SNPsites.html">SNPsites</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotMutSites.html">plotMutSites</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/find_SNP-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="va">snps</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/find_SNP-2.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="va">snps</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="sitePath_files/figure-html/find_SNP-3.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2022-01-14 r81484)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] sitePath_1.11.0  BiocStyle_2.23.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] ggrepel_0.9.1       Rcpp_1.0.8          ape_5.6-1          </span>
<span class="co">#&gt;  [4] lattice_0.20-45     tidyr_1.1.4         rprojroot_2.0.2    </span>
<span class="co">#&gt;  [7] digest_0.6.29       utf8_1.2.2          R6_2.5.1           </span>
<span class="co">#&gt; [10] evaluate_0.14       highr_0.9           ggplot2_3.3.5      </span>
<span class="co">#&gt; [13] pillar_1.6.4        ggfun_0.0.4         yulab.utils_0.0.4  </span>
<span class="co">#&gt; [16] rlang_0.4.12        lazyeval_0.2.2      jquerylib_0.1.4    </span>
<span class="co">#&gt; [19] rmarkdown_2.11      pkgdown_2.0.2.9000  labeling_0.4.2     </span>
<span class="co">#&gt; [22] textshaping_0.3.6   desc_1.4.0          stringr_1.4.0      </span>
<span class="co">#&gt; [25] munsell_0.5.0       compiler_4.2.0      xfun_0.29          </span>
<span class="co">#&gt; [28] pkgconfig_2.0.3     systemfonts_1.0.3   gridGraphics_0.5-1 </span>
<span class="co">#&gt; [31] htmltools_0.5.2     tidyselect_1.1.1    tibble_3.1.6       </span>
<span class="co">#&gt; [34] gridExtra_2.3       bookdown_0.24       fansi_1.0.2        </span>
<span class="co">#&gt; [37] crayon_1.4.2        dplyr_1.0.7         MASS_7.3-55        </span>
<span class="co">#&gt; [40] grid_4.2.0          nlme_3.1-155        jsonlite_1.7.3     </span>
<span class="co">#&gt; [43] gtable_0.3.0        lifecycle_1.0.1     magrittr_2.0.1     </span>
<span class="co">#&gt; [46] scales_1.1.1        tidytree_0.3.7      stringi_1.7.6      </span>
<span class="co">#&gt; [49] cachem_1.0.6        farver_2.1.0        fs_1.5.2           </span>
<span class="co">#&gt; [52] ggtree_3.3.1        seqinr_4.2-8        bslib_0.3.1        </span>
<span class="co">#&gt; [55] ellipsis_0.3.2      ragg_1.2.1          generics_0.1.1     </span>
<span class="co">#&gt; [58] vctrs_0.3.8         RColorBrewer_1.1-2  tools_4.2.0        </span>
<span class="co">#&gt; [61] treeio_1.19.1       ade4_1.7-18         ggplotify_0.1.0    </span>
<span class="co">#&gt; [64] glue_1.6.0          purrr_0.3.4         parallel_4.2.0     </span>
<span class="co">#&gt; [67] fastmap_1.1.0       yaml_2.2.1          colorspace_2.0-2   </span>
<span class="co">#&gt; [70] BiocManager_1.30.16 aplot_0.1.2         memoise_2.0.1      </span>
<span class="co">#&gt; [73] knitr_1.37          patchwork_1.1.1     sass_0.4.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Chengyang Ji.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
